### ISSUE TITLE
services bgp announcements are not restored 

### ISSUE THREAD

**Issue by nathanael-nsone** [metallb/metallb#1613]
I'm working in an environment with the following:

k3s version v1.24.3+k3s1 (990ba0e8)
go version go1.18.1
metallb installed via helm (chart version: metallb-0.13.4), app version (v0.13.4)
I'm using calico CNI but not using bird to establish bgp sessions
I have an nginx service running with the following config

> apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  selector:
    matchLabels:
      app: nginx
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:1
        ports:
        - name: http
          containerPort: 80

> ---
> apiVersion: v1
kind: Service
metadata:
  name: nginx
  annotations:
    metallb.universe.tf/address-pool: pocv4public
    metallb.universe.tf/loadBalancerIPs: _scrubbed_200.218,_scrubbed-v6_:ff::1
    #    metallb.universe.tf/loadBalancerIPs: 8.8.8.8,2001:4860:4860::8888
spec:
  ipFamilyPolicy: PreferDualStack
  ipFamilies:
    - IPv6
    - IPv4
  ports:
>   - name: http
    port: 80
    protocol: TCP
    targetPort: 80
  selector:
    app: nginx
  type: LoadBalancer
  externalTrafficPolicy: Local

I'm using BGP mode.
I've noticed that sometimes after the external IP announcement is withdrawn because the pods are cycled due to changes, once the service pods are back again the BGP announcements are not restored.

Attached are two files with more details
[bug-report.txt](https://github.com/metallb/metallb/files/9577297/bug-report.txt)
[metallb-report.txt](https://github.com/metallb/metallb/files/9577298/metallb-report.txt)

**Comment by fedepaol** [comment]
Hi, sorry for the delay in replying.
What I can see on the logs is (assuming I am interpreting correctly the obfuscated values):

The service has ExternalTrafficPolicy = Local (meaning, only the speakers running on nodes with an endpoint will announce).

When everything is working, the endpoint is:
```
{"caller":"service_controller.go:122","controller":"ServiceReconciler","enqueueing":"default/nginx","epslice":"{\"metadata\":{\"name\":\"nginx-kkln9\",\"generateName\":\"nginx-\",\"namespace\":\"default\",\"uid\":\"f086f6ab-1ab3-4488-b73d-d9fb1a70505f\",\"resourceVersion\":\"887428\",\"generation\":79,\"creationTimestamp\":\"2022-08-01T15:41:01Z\",\"labels\":{\"endpointslice.kubernetes.io/managed-by\":\"endpointslice-controller.k8s.io\",\"kubernetes.io/service-name\":\"nginx\"},\"annotations\":{\"endpoints.kubernetes.io/last-change-trigger-time\":\"2022-09-15T15:46:28Z\"},\"ownerReferences\":[{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"name\":\"nginx\",\"uid\":\"d2c02cef-2afb-4e32-ac88-986c953c2b3e\",\"controller\":true,\"blockOwnerDeletion\":true}],\"managedFields\":[{\"manager\":\"k3s\",\"operation\":\"Update\",\"apiVersion\":\"discovery.k8s.io/v1\",\"time\":\"2022-09-15T15:46:28Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:addressType\":{},\"f:endpoints\":{},\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:endpoints.kubernetes.io/last-change-trigger-time\":{}},\"f:generateName\":{},\"f:labels\":{\".\":{},\"f:endpointslice.kubernetes.io/managed-by\":{},\"f:kubernetes.io/service-name\":{}},\"f:ownerReferences\":{\".\":{},\"k:{\\\"uid\\\":\\\"d2c02cef-2afb-4e32-ac88-986c953c2b3e\\\"}\":{}}},\"f:ports\":{}}}]},\"addressType\":\"IPv4\",\"endpoints\":[{\"addresses\":[\"10.43.0.135\"],\"conditions\":{\"ready\":true,\"serving\":true,\"terminating\":false},\"targetRef\":{\"kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"nginx-595cfdf858-xrzxv\",\"uid\":\"a1b458cc-0dea-497a-944e-47be1769adf9\"},\"nodeName\":\"scrubbed-host\"},{\"addresses\":[\"10.43.0.207\"],\"conditions\":{\"ready\":true,\"serving\":true,\"terminating\":false},\"targetRef\":{\"kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"nginx-595cfdf858-nqdgr\",\"uid\":\"9921684c-f5e8-48d7-a9d7-c377e34b14ab\"},\"nodeName\":\"scrubbed-host.fakezone.co\"}],\"ports\":[{\"name\":\"http\",\"protocol\":\"TCP\",\"port\":80}]}","level":"debug","ts":"2022-09-15T15:58:16Z"}
```

TL;DR of above, we have two endpoints, one on **\"nodeName\":\"scrubbed-host\"** and the other on **nodeName\":\"scrubbed-host.fakezone.co\"**}]


When we have one single endpoint though:


```json
{"caller":"service_controller.go:122","controller":"ServiceReconciler","enqueueing":"default/nginx","epslice":"{\"metadata\":{\"name\":\"nginx-kkln9\",\"generateName\":\"nginx-\",\"namespace\":\"default\",\"uid\":\"f086f6ab-1ab3-4488-b73d-d9fb1a70505f\",\"resourceVersion\":\"888179\",\"generation\":89,\"creationTimestamp\":\"2022-08-01T15:41:01Z\",\"labels\":{\"endpointslice.kubernetes.io/managed-by\":\"endpointslice-controller.k8s.io\",\"kubernetes.io/service-name\":\"nginx\"},\"annotations\":{\"endpoints.kubernetes.io/last-change-trigger-time\":\"2022-09-15T16:10:16Z\"},\"ownerReferences\":[{\"apiVersion\":\"v1\",\"kind\":\"Service\",\"name\":\"nginx\",\"uid\":\"d2c02cef-2afb-4e32-ac88-986c953c2b3e\",\"controller\":true,\"blockOwnerDeletion\":true}],\"managedFields\":[{\"manager\":\"k3s\",\"operation\":\"Update\",\"apiVersion\":\"discovery.k8s.io/v1\",\"time\":\"2022-09-15T16:10:16Z\",\"fieldsType\":\"FieldsV1\",\"fieldsV1\":{\"f:addressType\":{},\"f:endpoints\":{},\"f:metadata\":{\"f:annotations\":{\".\":{},\"f:endpoints.kubernetes.io/last-change-trigger-time\":{}},\"f:generateName\":{},\"f:labels\":{\".\":{},\"f:endpointslice.kubernetes.io/managed-by\":{},\"f:kubernetes.io/service-name\":{}},\"f:ownerReferences\":{\".\":{},\"k:{\\\"uid\\\":\\\"d2c02cef-2afb-4e32-ac88-986c953c2b3e\\\"}\":{}}},\"f:ports\":{}}}]},\"addressType\":\"IPv4\",\"endpoints\":[{\"addresses\":[\"10.43.0.209\"],\"conditions\":{\"ready\":true,\"serving\":true,\"terminating\":false},\"targetRef\":{\"kind\":\"Pod\",\"namespace\":\"default\",\"name\":\"nginx-595cfdf858-kj5w2\",\"uid\":\"6543dda4-47f2-4785-9cf7-261cc5a4e7d6\"},\"nodeName\":\"scrubbed-host.fakezone.co\"}],\"ports\":[{\"name\":\"http\",\"protocol\":\"TCP\",\"port\":80}]}","level":"debug","ts":"2022-09-15T16:10:16Z"}
```

Here we have one single endpoint running on **\"nodeName\":\"scrubbed-host.fakezone.co\"}**. Now, in your configuration you are peering only from "scrubbed-host"

From the BGPPeer:
**nodeSelectors\\\":[{\\\"matchLabels\\\":{\\\"kubernetes.io/hostname\\\":\\\"scrubbed-host\\\"}**, so only the speaker running on that node (likely, the one you sent the logs for) would be able to announce.

In case of external traffic policy = Local, each speaker matches the nodeName in the endpointslice with the node it's currently running on, and only if there's a running and healthy end point it announces the service.

Can you check if this was the case? Another test can be changing the external traffic policy to "cluster".

I am going to close the issue because everything is suggesting that is the cause of the problem, but if your endpoint is still running on the right host or I missed something feel free to expand and reopen the issue.
